### PR TITLE
[RISCV][MC] MC layer support for the experimental zalasr extension

### PR DESCRIPTION
**Pull request by mehnadnerd**
This PR implements the RISC-V Atomic Load-Acquire and Store-Release Extension (Zalasr).
See also https://github.com/mehnadnerd/riscv-zalasr.

### CODE REVIEW

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
Please add some regression tests:
+ clang/test/Preprocessor/riscv-target-features.c
+ llvm/test/CodeGen/RISCV/attributes.ll
+ llvm/test/MC/RISCV/attribute-arch.s
+ llvm/test/MC/RISCV/zalasr-valid.s
+ llvm/test/MC/RISCV/zalasr-invalid.s

See also https://reviews.llvm.org/D149248.

**Comment by asb** [comment]
Zalasr is using the namespace reserved for standard extensions, but doesn't seem to be being developed under github.com/riscv - why is that? Is the repo going to get moved to that org? I see it _is_ an official in-development extension based on your riscv-opcodes PR, but it would be great to remove that ambiguity.

**Comment by wangpc-pp** [comment]
If we are going to support `Zalasr`, can we consider #67108 too?

**Comment by mehnadnerd** [comment]
> Zalasr is using the namespace reserved for standard extensions, but doesn't seem to be being developed under github.com/riscv - why is that? Is the repo going to get moved to that org? I see it _is_ an official in-development extension based on your riscv-opcodes PR, but it would be great to remove that ambiguity.
I'm honestly not entirely sure--I think it will be moved into there eventually, but I am unsure of the process.

**Comment by mehnadnerd** [comment]
I've updated it to add regression tests.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -20506,6 +20506,15 @@ unsigned RISCVTargetLowering::getCustomCtpopCost(EVT VT,
   return isCtpopFast(VT) ? 0 : 1;
 }
 
+bool RISCVTargetLowering::shouldInsertFencesForAtomic(
+    const Instruction *I) const {
+  if (Subtarget.hasStdExtZalasr()) {
+    return false;
+  } else {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:20513]: No else after return

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td [hunk 2]
```diff
@@ -105,22 +105,63 @@ defm AMOMAXU_D  : AMO_rr_aq_rl<0b11100, 0b011, "amomaxu.d">,
...
+}
+
 // Atomic load/store are available under both +a and +force-atomics.
-// Fences will be inserted for atomic load/stores according to the logic in
-// RISCVTargetLowering::{emitLeadingFence,emitTrailingFence}.
 let Predicates = [HasAtomicLdSt] in {
-  def : LdPat<atomic_load_8,  LB>;
-  def : LdPat<atomic_load_16, LH>;
-  def : LdPat<atomic_load_32, LW>;
+  def : LdPat<relaxed_load<atomic_load_8>,  LB>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoA.td:153]: If this is an MC patch as the title says, we shouldn't be touching any isel patterns.

**Comment by mehnadnerd** [comment]
Still needs tests for the lowering.

File: clang/test/Preprocessor/riscv-target-features.c [hunk 3]
```diff
@@ -1120,6 +1121,14 @@
 // RUN:   -o - | FileCheck --check-prefix=CHECK-SMEPMP-EXT %s
 // CHECK-SMEPMP-EXT: __riscv_smepmp  1000000{{$}}
 
+// RUN: %clang --target=riscv32 -menable-experimental-extensions \
+// RUN: -march=rv32i_zalasr1p0 -x c -E -dM %s \
```
Comment (topperc) [inline clang/test/Preprocessor/riscv-target-features.c:1125]: Indent line continuations 2 spaces and drop `-x c`. This was recently updated in the reset of the file.

File: llvm/docs/RISCVUsage.rst [hunk 4]
```diff
@@ -228,6 +228,9 @@ The primary goal of experimental support is to assist in the process of ratifica
 ``experimental-zacas``
   LLVM implements the `1.0-rc1 draft specification <https://github.com/riscv/riscv-zacas/releases/tag/v1.0-rc1>`_.
 
+``experimental-zalasr``
+  LLVM implements the `most recent specification <https://github.com/mehnadnerd/riscv-zalasr>`_.
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:232]: Can we nail this down to a version or a date? "most recent" is wrong if that repo changes.

File: llvm/lib/Target/RISCV/RISCVFeatures.td [hunk 5]
```diff
@@ -797,6 +797,13 @@ def FeatureStdExtSvpbmt
     : SubtargetFeature<"svpbmt", "HasStdExtSvpbmt", "true",
                        "'Svpbmt' (Page-Based Memory Types)">;
 
+def FeatureStdExtZalasr
+    : SubtargetFeature<"experimental-zalasr", "HasStdExtZalasr", "true",
+                       "'Zalasr' (Load-Acquire and Store-Release Instructions)">;
+def HasStdExtZalasr : Predicate<"Subtarget->hasStdExtZalasr()">,
+                               AssemblerPredicate<(all_of FeatureStdExtZalasr),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFeatures.td:804]: Align `AssemblerPredicate` with `Predicate` on the previous line. This file was recently reformatted.

File: llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+let Predicates = [HasStdExtZalasr, IsRV64] in {
+defm LD_AQ : LAQ_r_aq_rl<0b011, "ld">;
+defm SD_RL : SRL_r_aq_rl<0b011, "sd">;
+} // Predicates = [HasStdExtZalasr, IsRV64]
+
+//===----------------------------------------------------------------------===//
+// Pseudo-instructions and codegen patterns
+//===----------------------------------------------------------------------===//
+
+class PatLAQ<SDPatternOperator OpNode, RVInst Inst, ValueType vt = XLenVT>
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoZalasr.td:66]: Don't add patterns in MC patch.

File: llvm/lib/Target/RISCV/RISCVInstrInfoA.td (hunk 2 above)
Comment (mehnadnerd) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:153]: This was originally an MC patch but I am adding isel support as well. What should I change the title to?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:153]: We usually want MC and CodeGen changes as separate PRs.

Comment (mehnadnerd) [reply llvm/lib/Target/RISCV/RISCVInstrInfoA.td:153]: Oh. I will separate it out then.

**Comment by mehnadnerd** [comment]
Closing because superseded by https://github.com/llvm/llvm-project/pull/79911.

**Pull request closed without merging**